### PR TITLE
Ionic 2 Support

### PR DESCRIPTION
**Pull request by ericb**
### WORK IN PROGRESS

- [x] bulk source conversion to Typescript
- [x] build ES6 distribution files
- [x] Implement new configuration setup per #41 
- [x] Do #94
- [ ] Regression testing

### CODE REVIEW

**Comment by danielgolub** [comment]
Hey guys,
Any updates with this issue?

**Comment by nunoarruda** [comment]
I want to test this. How do I install it in a Ionic 2 project? The `ionic add ionic-platform-web-client` command still doesn't support v2 apps.

**Comment by ericb** [comment]
@danielgolub As far as I know the branch is in good shape -- at this point it's testing / major documentation updates.

@nunoarruda You can test this by cloning the repo and running an `npm install ~/path/to/cloned/project`, but there is no documentation provided for Ionic 2 usage just yet, but the code is all available. you could easily look at the src for class names... everything is exported under the `ionic-platform-web-client` module.

**Comment by anaganisk** [comment]
@ericb I've tried to npm install from cloned repo but it does'nt work an I get the following error
`Error TS2307: Cannot find module 'ionic-platform-web-client'.`

**Comment by anaganisk** [comment]
@ericb  I've tried to npm install from cloned repo but it does'nt work and I get the following error
`Error TS2307: Cannot find module 'ionic-platform-web-client'.` Is there something I' missing? BTW I've clone ionic2-support branch and also tried with the master branch too

**Comment by ericb** [comment]
@anaganisk Are you attempting this in a typescript based Ionic2 project, or just ES6?

**Comment by anaganisk** [comment]
@ericb Its Typescript based ionic2

**Comment by ericb** [comment]
@anaganisk Alright, I'll try out a typescript based project. I'm likely just missing something specific for a typescript project.

**Comment by anaganisk** [comment]
@ericb thanks.

**Comment by Maistho** [comment]
So I have updated to using this branch in my (Ionic 1.3) application, and have not found any issues so far.
What I have tested and found to be working:
- Initialisation (using Ionic.Core.init)
- Push register and unregister
- Push receiving
- User login / logout

A nice side-effect of the new initialisation is that I can set dev_push to true/false depending on if we're running in Chrome or on a device (currently checking for window.cordova), which was a gripe for our team previously. :tada:

The documentation for initialising needs to be updated (of course), but in general it seems to be working as expected.

**Comment by anaganisk** [comment]
@Maistho This thread is related to the support of Ionic2 and I dont think Ionic ver1.3 is relevant here at all. And ofcourse it will work in ionic 1.x since, things will be added to support Ionic 2, most probablyz. Please correct me if I am wrong.

**Comment by Maistho** [comment]
@anaganisk Well, it could also have been broken, as the current implementation is (latest release gives a faulty promise result from unregistering a push token). 
And as most people will probably use ionic 1.x for a long time to come, it seems reasonable to make sure it works for that too. :)

**Comment by ericb** [comment]
@Maistho Huge thanks for running some tests! One clarification though, are you saying the currently released version of the lib gives a faulty promise, or this branch? I believe their was an issue with the http lib I was bundling previously which did not correctly handle 204 status codes, but this branch should have that fixed up. Also, how did you find the new core configuration with init? I'm hoping it alleviates all the issues people were having with the injected config.

@anaganisk This is relevant for 1.x and 2.x versions of Ionic. One of the big jumps here was the move to typescript for the main source (which effects the builds of the ES6/ES5 distribution files).

**Comment by Maistho** [comment]
@ericb To clarify: the current release (master / 0.7) has issues with 204 status codes. This branch (ionic2-support) does not, which is great! =) 

I think the new configuration is much better. Previously I had a gulp task which searched for some string (IONIC_CONFIG etc) and replaced it when building, and removing that from my build felt great. Creating the settings at runtime is also since I have more freedom. 

Things I wondered about:
- Can settings be changed after initialisation? This could be useful in some cases. 
- Is it possible to set the settings from within Angular using a provider? 

In general the documentation for using the angular wrapper services could be improved, but maybe I'm supposed to just use the "raw" objects instead?

**Comment by anaganisk** [comment]
@ericb @Maistho Thank you, for clarifying. I was a bit confused.

**Comment by dwieeb** [comment]
The latest version of the cli by default starts Ionic apps with browserify and a ES2015 config from these [gulp tasks](https://github.com/driftyco/ionic-gulp-tasks). As far as I'm aware, the current `ionic2-support` branch should work perfectly if your app imports `ionic-platform-web-client` from ES6, but TypeScript support is currently broken. We should be pushing some updates to the branch this week to address it.

**Comment by cklanac** [comment]
ES6 and TypeScript projects work after manual build...

Initially i ran `ionic upload` on a new, clean Ionic 2 project and got the following warning... "WARN: No 'upload:before' gulp task found! If your app requires a build step, you may want to ensure it runs before upload." So I ran `ionic build` and re-uploaded and it worked perfectly. I don't think Ionic v1 projects require a build before uploading, right?

Anyway, anyone who's trying this out, this worked for me...
```
# Clone ionic-platform-web-client and checkout ionic2-support branch
git clone -b ionic2-support https://github.com/driftyco/ionic-platform-web-client.git

# Create ionic2 TypeScript project
ionic start Ionic2TS-Platform-Test --v2 --ts
cd Ionic2TS-Platform-Test
ionic build

# Install platform from clone, login, init and upload
npm install ../ionic-platform-web-client
ionic login
ionic io init
ionic upload
```

**Comment by Manduro** [comment]
@dwieeb @ericb I've been testing this in a TypeScript project. Some things to report:

**Injection errors**
When adding Push or Analytics to @App providers:
```reflective_provider.js:182Uncaught Cannot resolve all parameters for 'Push'(?). Make sure that all the parameters are decorated with Inject or have valid type annotations and that 'Push' is decorated with Injectable.```

**Two-way binding template trouble**
When adding the working Deploy to @App providers, I get a successful `Ionic Deploy: init` log. Something then goes wrong with template bindings, which is odd as the platform client should not interfere with that. An example: When I open a tab, it defaults into loading state until some data is loaded. But even when the data has been loaded, the tab stays in it's loading state and it's not update until I have some interaction with the app (e.g. click the tab again or some navbar button or anything really).

**Comment by dwieeb** [comment]
@Manduro By saying "`@App` providers" I assume you mean to your `@Component` annotation you added the Push class to `providers: [...]`. If so, that is not how we intend this library to be used. This is not an angular 1 nor angular 2 library. While we hope to add stuff _on top_ to integrate with angular 2's dependency injection, we're not quite there yet. Please give it a try instantiating objects normally.

**Comment by Manduro** [comment]
@dwieeb That makes sense, thanks. The binding trouble is still there though!

**Comment by dwieeb** [comment]
@Manduro Hmm. I assumed that was also a result of the class being listed in your `providers`. Could you possibly post your code in a gist?

**Comment by anaganisk** [comment]
@dwieeb Is typescript support working?

**Comment by Manduro** [comment]
@dwieeb It is unrelated to my code. It happens even when just including it in my build and doing nothing with it.
```javascript
import { IonicPlatform } from 'ionic-platform-web-client';
console.log(IonicPlatform);
```
Without it, everything is fine.

**Comment by dwieeb** [comment]
Folks, I'm looking for people to start testing v2 support for our platform client for the latest Ionic v2 (beta 8). 

The biggest change is that instead of installing this package directly, most people will want to use the new [ionic-platform-web-client-angular](https://github.com/driftyco/ionic-platform-web-client-angular#ionic-platform-web-client-angular) package (we're working on the name 😛). This package is basically a thin wrapper for Angular 2 support in Ionic 2. The README should be enough to get you going.

Here's an example of the config you would pass into that `providers` function:
```javascript
...

let config = {
  'core': {
    'app_id': "abcd123",
    'gcm_key': "123456789" // We'd like to move this to 'push' soon
  },
  'push': {
    'onNotification': (notification) => {
      let payload = notification.payload;
      console.log(notification, payload);
    }
  }
};

ionicBootstrap(MyApp, providers(config));
```
Basically, `'core'` is what `IonicPlatform.init(coreConfig)` was and `'push'` is what `new Push(pushConfig)` was.

This is a very early package and there's more we can do, like export types such as `PushMessage`.

Let me know what issues are encountered, things to be improved.

**Comment by anaganisk** [comment]
@dwieeb I am in

**Comment by Manduro** [comment]
@dwieeb Thanks for the beta package release! The trouble I had before is still there when using the angular wrapper. This bug makes apps using the platform-client essentially unusable. Here is a plunker demonstrating the bug: http://plnkr.co/edit/38drD3Af8Xmi567VWRhA?p=preview

**Comment by dwieeb** [comment]
@Manduro I see! Sorry. We were including the `es6-promise` shim where we really only needed to include the typings. Here's your plunker using the new angular wrapper and a configured systemjs: http://plnkr.co/edit/JEIb7KUjId3WrwmuK6kE?p=preview Thank you for the example!

**Comment by Manduro** [comment]
@dwieeb Thank you for looking into it! It works here and when the npm cdn gets updated that plunker you mention will probably start working as well ;)

**Comment by dwieeb** [comment]
@Manduro Nope, that's just me not knowing how to use plunker yet. Here you go: http://plnkr.co/edit/mMuaKInP0wTMjUnt5bbn?p=preview

**Comment by dwieeb** [comment]
Update: We created a new repo for v2 support: https://github.com/driftyco/ionic-cloud We did this because bower looks at `master` on Github when it installs a package, and we didn't want to break existing installations with v1 users. The installation/setup steps are most up-to-date here: https://github.com/driftyco/ionic-cloud-angular#readme

**Comment by joshgarwood** [comment]
Thanks @dwieeb !  In running ```npm install --save @ionic/cloud-angular``` I get an error about @ionic/cloud-angular not being in the npm registry...  Should I be getting @ionic/cloud instead?  I tried that, but the example in the ionic-cloud-angular documentation didn't work :/

**Comment by dwieeb** [comment]
@joshgarwood Try now?

**Comment by dwieeb** [comment]
Closing this pull request. Angular 2 & Ionic 2 stuff will be in [this repo](https://github.com/driftyco/ionic-cloud-angular). I'm [looking for feedback](https://github.com/driftyco/ionic-cloud-angular/issues/1)!
